h2/src/main/org/h2/value/ValueJson.java

```diff
@@ -0,0 +1,180 @@
+    }
+
+    @Override
+    public int getValueType() {
+        return JSON;
+    }
+
+    @Override
+    public String getString() {
+        return string;
+    }
+
+    @Override
+    public JsonNode getObject() {
+        return json;
+    }
+    
+    /*
+     * approximate evaluation
+     * At least 2 bytes per char symbol and 6 bytes per binary object 
+     * Plus 88+40 bytes on empty value
+     */
+    @Override
+    public int getMemory() {
+        return binarySize * 8 + 128;
+    }
+
+    @Override
+    public void set(PreparedStatement prep, int parameterIndex) throws SQLException {
+        // TODO Auto-generated method stub
+
+    }
+
+    @Override
+    public int hashCode() {
+        return json.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (other instanceof ValueJson) {
+            return this.json.equals(((ValueJson) other).getObject());
+        } else if (other instanceof JsonNode) {
+            return this.json.equals(other);
+        }
+        return false;
+    }
+
+    @Override
+    public Value subtract(Value v) {
+        JsonNode json = this.json;
+        String toSub = v.getString();
+        if (json.isObject() && json.has(toSub)) {
+            /* Used copy to save original object */
+            ObjectNode result = json.deepCopy();
+            String key = v.getString();
+            result.remove(key);
+            return get(result);
+        } else if (json.isArray()) {
+            ArrayNode result = mapper.createArrayNode();
+            String value = v.getString();
+            for (JsonNode node : json) {
+                if (!node.asText().equals(value)) {
+                    result.add(node);
+                }
+            }
+            return get(result);
+        } else {
+            return this;
+        }
+    }
+    
+    @Override
+    public Value add(Value v) {
+        JsonNode l = this.json.deepCopy();
+        JsonNode r = ((ValueJson) v).getObject();
+        if (l.isArray()) {
+            ArrayNode lA = (ArrayNode) l;
+            if (r.isArray()) {
+                ArrayNode rA = (ArrayNode) r;
+                lA.addAll(rA);
+            } else {
+                lA.add(r);
+            }
+            return get(lA);
+        } else if (l.isObject() && r.isObject()){
+           ObjectNode lO = (ObjectNode) l;
+           ObjectNode rO = (ObjectNode) r;
+           lO.putAll(rO);
+           return get(lO);
+        } else {
+            ArrayNode result = mapper.createArrayNode();
+            result.add(l).add(r);
+            return get(result);
+        }
+            
+    }
+
+    @Override
+    public int compareTypeSafe(Value v, CompareMode mode) {
+        // TODO Auto-generated method stub
+        return 0;
+    }
+
+    @Override
+    public StringBuilder getSQL(StringBuilder builder) {
+        return builder.append(getString()).append("::JSON");
+    }
+    
+    @Override
+    public String toString() {
```

Comment:
JSON can be very large, it's better not to override this method here, it can create problems for debuggers. Implementation in a superclass truncates SQL to reasonable length.

Reply:
Fixed [here](https://github.com/lazarevnik/h2database/commit/004d312ff1cce4325f45302e3320e8450f17eacc)